### PR TITLE
[SPARK-39663][SQL][TESTS] Add UT for MysqlDialect listIndexes method

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Add complemented UT for MysqlDialect's lustIndexes method.

### Why are the changes needed?
Add UT for existed function & improve test coverage.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### CODE REVIEW

**Comment by panbingkun** [comment]
@huaxingao 
The pr related to: https://github.com/apache/spark/pull/37031

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 1]
```diff
@@ -219,11 +221,19 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
...
+          assert(indexes(0).indexName() == "i1")
+        }
+
         sql(s"CREATE index i2 ON $catalogName.new_table (col2, col3, col5)" +
           s" OPTIONS ($indexOptions)")
-
-        assert(jdbcTable.indexExists("i1") == true)
         assert(jdbcTable.indexExists("i2") == true)
+        if (supportListIndexes) {
+          assert(jdbcTable.listIndexes().size == 2)
```
Comment (huaxingao) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:235]: nit: Could you verify the indexName too?

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 2]
```diff
@@ -219,11 +221,19 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
           s" The supported Index Types are:"))
 
         sql(s"CREATE index i1 ON $catalogName.new_table USING BTREE (col1)")
+        assert(jdbcTable.indexExists("i1") == true)
```
Comment (srowen) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:224]: Nit: drop `== true`

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala (hunk 1 above)
Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala]: Updated

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala (hunk 2 above)
Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala]: Updated

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 3]
```diff
@@ -219,11 +221,21 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
...
         sql(s"CREATE index i2 ON $catalogName.new_table (col2, col3, col5)" +
           s" OPTIONS ($indexOptions)")
-
-        assert(jdbcTable.indexExists("i1") == true)
-        assert(jdbcTable.indexExists("i2") == true)
+        assert(jdbcTable.indexExists("i2"))
+        if (supportListIndexes) {
+          val indexes = jdbcTable.listIndexes()
+          assert(indexes.size == 2)
+          assert(indexes.map(index => index.indexName()).sortWith(_ < _) === Array("i1", "i2"))
```
Comment (LuciferYang) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:237]: How about using `sorted`?

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 4]
```diff
@@ -234,10 +246,18 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
...
+        if (supportListIndexes) {
+          val indexes = jdbcTable.listIndexes()
+          assert(indexes.size == 1)
+          assert(indexes(0).indexName() == "i2")
+        }
+
+        sql(s"DROP index i2 ON $catalogName.new_table")
         assert(jdbcTable.indexExists("i2") == false)
+        if (supportListIndexes) {
+          assert(jdbcTable.listIndexes().size == 0)
```
Comment (LuciferYang) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:259]: how about `assert(jdbcTable.listIndexes().isEmpty)`?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 5]
```diff
@@ -206,7 +206,7 @@ private case object MySQLDialect extends JdbcDialect with SQLConfHelper {
           val indexName = rs.getString("key_name")
           val colName = rs.getString("column_name")
           val indexType = rs.getString("index_type")
-          val indexComment = rs.getString("Index_comment")
+          val indexComment = rs.getString("index_comment")
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:209]: Can't get correct results with `Index_comment`? This change looks more like a bug fix?

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:209]: The column name is actually `Index_comment`, but case doesn't matter. In Java [doc](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html), it has "Column names used as input to getter methods are case insensitive".

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala (hunk 3 above)
Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:237]: if not use,maybe it return Array("i1", "i2") or Array("i12, "i1")

Comment (LuciferYang) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:237]: I mean using `indexes.map(index => index.indexName()). sorted `, but it doesn't matter

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala (hunk 5 above)
Comment (panbingkun) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:209]: eg
mysql> SHOW INDEXES FROM xxx;
+----------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| Table    | Non_unique | Key_name | Seq_in_index | Column_name | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment |
+----------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+
| xxx |          0 | PRIMARY  |            1 | jobId       | A         |           0 |     NULL | NULL   |      | BTREE      |         |               |
+----------+------------+----------+--------------+-------------+-----------+-------------+----------+--------+------+------------+---------+---------------+

Comment (panbingkun) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:209]: maybe consistency seems more natural, haha

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala (hunk 4 above)
Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:259]: Updated

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 6]
```diff
@@ -219,11 +221,21 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
           s" The supported Index Types are:"))
 
         sql(s"CREATE index i1 ON $catalogName.new_table USING BTREE (col1)")
+        assert(jdbcTable.indexExists("i1"))
+        if (supportListIndexes) {
+          val indexes = jdbcTable.listIndexes()
+          assert(indexes.size == 1)
+          assert(indexes(0).indexName() == "i1")
```
Comment (srowen) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:228]: While we're nit-picking, `indexes.head`? doesn't matter

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala [hunk 7]
```diff
@@ -219,11 +221,21 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
...
         sql(s"CREATE index i2 ON $catalogName.new_table (col2, col3, col5)" +
           s" OPTIONS ($indexOptions)")
-
-        assert(jdbcTable.indexExists("i1") == true)
-        assert(jdbcTable.indexExists("i2") == true)
+        assert(jdbcTable.indexExists("i2"))
+        if (supportListIndexes) {
+          val indexes = jdbcTable.listIndexes()
+          assert(indexes.size == 2)
+          assert(indexes.map(index => index.indexName()).sorted === Array("i1", "i2"))
```
Comment (srowen) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:237]: `.map(_.indexName())`

**Review by LuciferYang — APPROVED** [review]
LGTM +1

**Review by huaxingao — APPROVED** [review]

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**